fix up python, matlab in tutorials/doc

jrscott opened the pull request:
## What changes does this PR introduce?
doc+python/matlab codes

jahn commented on verification/tutorial_barotropic_gyre/input/gendata.py:
```diff
@@ -0,0 +1,27 @@
+
+# Zonal wind-stress, located at u-points
+tauMax=0.; # wind stress maximum
+x=np.array(np.arange(nx)-1)/(nx-2) # x-coordinate, located at XG points
+y=np.array(np.arange(ny)-.5)/(ny-2) # y-coordinate, located at YC points
```
The np.array() is unnecessary as np.arange already returns a numpy array

rabernat commented:
I'm curious why you decided to remove xarray from the python tutorial.

jrscott commented:
still have more commits to make but @jahn see my new python routine that does not use xarray. Will add another python script that uses xarray.

jrscott commented:
@rabernat after surveying people who use python, there are some that are not familiar with xarray, and I wanted to keep this as simple as possible. Moreover I wanted to have the python and matlab version be fairly similar, i.e. "matlab-y" python code here. But as you probably saw I was going to include a second python version that did use xarray.

This is the first real tutorial showing how to look at output (BT gyre is completely trivial) and I wanted to show some basic loading and plotting from first principles. Plan is to graduate to xarray, xgcm and other tools in later tutorials, using them to facilitate some plots, pointing to these packages' external doc. I'm learning as I go along here -- but will use xarray at least in making a notebook for the python SO analysis (right now I just have a matlab script). Happy to here any thoughts or suggestions. Currently working on new (coarse) 2.8x2.8 lat-lon global and regional Arctic setups for tutorials; the latter especially (cubed sphere cap) will need some tools to help analyze.

jrscott commented:
Reviewers: guide to what's new/edited - 

- gendata.m and .py edited/added for BT and BC gyre tutorials

- matlab and python (w/ and w/o xarray) added/signif. revisions. I've assumed basic matlab and python (numpy, matplotlib) knowledge but added a lot of text in .m and .py meant as tutorial explanation how to load, calculate, and plot output (for python w/xarray, I don't assume user knows xarray a priori but is learning it on the side; I try to point out why it is helpful however rather than teach its syntax). Please check that I'm calculating and plotting things correct w.r.t. C-grid etc. Also please check my python code, hoping it follows standard syntax and readability guidelines etc. Also mentioned MITgcmutils mnc_files as alternative to gluemncbig in .py file (@jahn  assumed that braces expand dependency put in)

- commit 2e396f4fcecce121f65c780c9f6229eb6c43012b left many important issues unexplained  and I've tried to fix this up. See edits in text describing data.mnc, output file description, explaining mnc outdir behavior with re-runs and if using mpi. (commit also broke things, fixed, added line to link global stat-diags file to top directory with other global .nc files)

- output file info about stat-diags file was missing, added

- replotted/improved the baroclinic gyre figures (matching up with .m and .py generated figs better)

(this PR should be squash-merged)

jrscott commented:
posting this email exchange:

just a footnote about this statement:

# Need to rename the DataArray coordinate Zmd000015 to Z to get depths plotted correctly (as MITgcm RC, cell centers in vertical).
dynDiag.THETA[-1,:,14,:].rename({'Zmd000015':'Z'}).where(grid.HFacC[:,14,:]!=0).plot.contourf(levels=np.arange(0,30,.2),cmap='coolwarm')

I don’t think my comment totally explains this; it is only by including the .where() command using HFacC (that has dimension Z) that this plots with the true depths in Z (python quirk it seems; assign_coords works w/o .where() however, a trick Oliver showed me)

My question: is there a bonafide reason why the mnc (metadata) depth dimension is seemingly not done in more standard fashion/usefully? Could this be fixed in the code? Or should I post this question to mitgcm support or devel? (or make an issue?)

(response from @jahn ):
You can do this for an entire dataset like this:

 dynDiag = dynDiag.rename({'Zmd000015':'Z'}) 

(or append it to the open_dataset call)

Note that if 'levels' is set in data.diagnostics, you also have to do this:

 dynDiag['Z'] = grid['Z'][dynDiag.diag_levels.astype(int)-1]

Otherwise it will not be able to pick the correct levels out of grid and
other variables with a full 'Z' dimension.  It is in this case that there is
a good reason for having a different dimension name.  xarray can, however,
handle this correctly as long as coordinate (in this case depth) values
are present in all variables.  It magically combines variables with a different
number of levels by keeping only those present in both!

-----------------------
Thanks @jahn 

If I understand correctly the issue with a diagnostics output list (file) consisting of 3D fields is that you can set it to produce just a subset of levels, ergo we have this strange Z dimension used. Your fix does work for my 3D set (where I don't specify levels) and my pref would be to so the rename step when I read in the variable (this allows me to eliminate the mysterious bits later on).  Perhaps I can add your second statement about in comments at the end of the .py file.

I'm a little surprised this Z dim issue has not been raised/made aware before, but it least in our inner MIT circle none of us use netcdf... but I know that others out there do. 

Also I didn't ask about the stat diags Z dim, but this seems the same issue (mixing surface, depth-integrated, and by-level output) ergo the Z variable is Zmd000015, again not the expected Z, but it is easier to work around this oddness here at least.

jrscott commented:
@jahn made an earlier pass through the python syntax (@jahn you probably should scan through new commit).
Also need someone to read comments and examine how I computed things (@jm-c you often have opinion on these things).
Also, looking for any additional concrete suggestions to make it easier to follow and read.
Otherwise I'll proceed to reformat the xarray version and then revisit the matlab version.

Final note, my intent was just providing scripts here, I had not planned to make jupyter notebooks for the python (but I guess I could...). I've done a notebook for the SO tutorial python script.

jrscott commented:
Reviewers: guide to what's new/edited -

- gendata.m and .py edited/added for BT and BC gyre tutorials
- doc section 3.9 redone, using snippet of genmake from BT gyre

- BC gyre matlab and python (w/ and w/o xarray) added/signif. revisions. I've assumed basic matlab and python (numpy, matplotlib) knowledge but added a lot of text in .m and .py meant as tutorial explanation how to load, calculate, and plot output (for python w/xarray, I don't assume user knows xarray a priori but is learning it on the side; I try to point out why it is helpful however rather than teach its syntax). Please check that I'm calculating and plotting things correct w.r.t. C-grid etc. Also mentioned MITgcmutils mnc_files as alternative to gluemncbig in .py file (@jahn assumed that braces expand dependency put in)

- commit 2e396f4 left many important issues unexplained and I've tried to fix this up. See edits in text describing data.mnc, output file description, explaining mnc outdir behavior with re-runs and if using mpi. (commit also broke scripts, fixed up, added line to link global stat-diags file to top directory with other global .nc files)

- output file info about stat-diags file was missing, added

- replotted/improved the baroclinic gyre figures (matching up with .m and .py generated figs better)

(this PR should be squash-merged)

jm-c commented on doc/examples/baroclinic_gyre/baroclinic_gyre.rst:
```diff
@@ -847,10 +860,18 @@ To accomplish this, we will make use of utility script :filelink:`utils/python/M
     % ./gluemncbig -o phiHydLow.nc mnc_test_*/phiHydLow*.t*.nc
+    % ln -s mnc_test_0001/dynStDiag.0000000000.t001.nc dynStDiag.nc
 
 For help using this utility, type ``gluemncbig --help``; note a python installation must for available for this script to work.
```
Something to clarify: the sentence above, starting by "note a python installation ..."  is not clear to me.

jrscott replied:
nothing mysterious here :), changed it that bit to: (note, this utility requires python)

jm-c commented on doc/examples/baroclinic_gyre/baroclinic_gyre.rst:
```diff
@@ -986,14 +1015,23 @@ Finally, to run the model (from your run directory), using four processes runnin
-Other than some additional `MPI <https://en.wikipedia.org/wiki/Message_Passing_Interface>`_-related information, the standard output content is identical to
-that from the single-process run.
+Other than some additional `MPI <https://en.wikipedia.org/wiki/Message_Passing_Interface>`_-related information,
+the standard output content is identical to that from the single-process run.
```
suggest to replace "identical" with "similar" (there are some differences, especially the grid stats report yC, dxG, dxC, rA ...)

jrscott replied:
changed indentical -> similar

jm-c reviewed (approved):
Plots looks good and all what I've seen seems good too. I left few minor comments (which are easy to address).

jm-c commented:
I forgot one thing: 
It would be nicer and should not be that hard to change these 2 tutorial ( barotropic & baroclinic ) README files to a README.md file.

jm-c commented:
All good.

jahn reviewed (approved):
LGTM
